src/tool/subcommands/state_migration_cmd.rs

```diff
@@ -1,140 +1,104 @@
-        tasks.push(tokio::spawn(async move {
-            download_bundle_if_needed(manifest, url)
-                .await
-                .with_context(|| format!("Failed to get {manifest}.car from {url}"))
-        }));
-    }
-
-    let mut car_streams = vec![];
-    for task in tasks {
-        let path = task.await??;
-        let car_stream = CarStream::new(tokio::io::BufReader::new(
-            tokio::fs::File::open(path).await?,
-        ))
-        .await?;
-        car_streams.push(car_stream);
-    }
-
-    let all_roots = car_streams
-        .iter()
-        .flat_map(|it| it.header.roots.iter())
-        .unique()
-        .cloned()
-        .collect::<Vec<_>>();
-
-    let car_writer = CarHeader::from(all_roots);
-
-    let mut zstd_encoder = ZstdEncoder::with_quality(
-        tokio::fs::File::create(Path::new(DEFAULT_BUNDLE_FILE_NAME)).await?,
-        async_compression::Level::Precise(zstd::zstd_safe::max_c_level()),
-    )
-    .compat_write();
-
-    car_writer
-        .write_stream_async(
-            &mut zstd_encoder,
-            &mut std::pin::pin!(merge_car_streams(car_streams).map(|b| {
-                let b = b.expect("There should be no invalid blocks");
-                (b.cid, b.data)
-            })),
+async fn generate_actor_bundle(output: PathBuf) -> anyhow::Result<()> {
+    let (roots, blocks) = stream::iter(ACTOR_BUNDLES.iter())
+        .then(
+            |ActorBundleInfo {
+                 manifest: root,
+                 url,
+             }| async move {
+                info!(%url, "downloading bundle");
+                let response = global_http_client()
+                    .get(url.clone())
+                    .send()
+                    .await?
+                    .error_for_status()?;
+                let bytes = response.bytes().await?;
+                let car = CarStream::new(Cursor::new(bytes)).await?;
+                ensure!(car.header.version == 1);
+                ensure!(car.header.roots.len() == 1);
+                ensure!(&car.header.roots[0] == root);
+                anyhow::Ok((root, car.try_collect::<Vec<_>>().await?))
+            },
         )
-        .await?;
-
-    Ok(())
-}
-
-async fn download_bundle_if_needed(root: &Cid, url: &Url) -> anyhow::Result<PathBuf> {
-    if !ACTOR_BUNDLE_CACHE_DIR.is_dir() {
-        fs::create_dir_all(ACTOR_BUNDLE_CACHE_DIR.as_path())?;
-    }
-    let cached_car_path = ACTOR_BUNDLE_CACHE_DIR.join(format!("{root}.car"));
-    if cached_car_path.is_file() {
-        if let Ok(file) = async_fs::File::open(&cached_car_path).await {
-            if let Ok(true) = is_bundle_valid(root, BufReader::new(file)).await {
-                return Ok(cached_car_path);
-            }
-        }
+        .try_collect::<Vec<_>>()
+        .await?
+        .into_iter()
+        .unzip::<_, _, Vec<&Cid>, Vec<_>>();
+
+    let roots = roots.into_iter().cloned().collect::<CidHashSet>();
+    let blocks = blocks.into_iter().flatten().try_fold(
```

Comment:
As per our discussion, let's sort the blocks such that the bundle is deterministic.